### PR TITLE
Trim empty frames in preview

### PR DESCRIPTION
**Pull request by codetheweb**
Closes https://github.com/sindresorhus/Gifski/issues/190.

Test file (has black frame in preview with current version of Gifski): 

https://user-images.githubusercontent.com/7410405/122309868-bcbfb100-ced4-11eb-9175-9b788dfe53f1.mp4

**Note**: this does not currently handle blank frames at the **end** of files. From my understanding, we would have to load all samples (`while let sampleBuffer = readerOutput.copyNextSampleBuffer()`) into memory to be able to iterate backwards through them. Happy to implement that as well if you think it's an acceptable trade-off.

### CODE REVIEW

File: Gifski/VideoValidator.swift [hunk 1]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
...
 
-		return .success(newAsset, newVideoMetadata)
+		// Find first non-empty frame
+		do {
+			let reader = try AVAssetReader(asset: newAsset)
+			let readerOutput = AVAssetReaderTrackOutput(track: newAsset.firstVideoTrack!, outputSettings: nil)
+			reader.add(readerOutput)
+			reader.startReading()
+
+			while let sampleBuffer = readerOutput.copyNextSampleBuffer() {
```
Comment (sindresorhus) [inline Gifski/VideoValidator.swift:194]: Per docs, you need to check the `.status` property, and handle the different cases properly.

**Comment by sindresorhus** [comment]
> From my understanding, we would have to load all samples (while let sampleBuffer = readerOutput.copyNextSampleBuffer()) into memory to be able to iterate backwards through them. Happy to implement that as well if you think it's an acceptable trade-off.

It's unclear to me whether or not this is a problem. From my understanding, the samples are not actually the decoded frames, but rather just metadata, so it should be pretty efficient. Needs to be tested though, but I would like to support trimming empty frames from the end too as that's more common than empty frames at the start.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: Gifski/VideoValidator.swift [hunk 2]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
...
+				// On first non-empty frame
+				if sampleBuffer.totalSampleSize != 0 {
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/VideoValidator.swift:196]: The `.isValid` property does [not check if the frame is empty](https://developer.apple.com/documentation/coremedia/1489151-cmsamplebufferisvalid) (tested and returns true when `totalSampleSize == 0`.

**Comment by codetheweb** [comment]
> It's unclear to me whether or not this is a problem. From my understanding, the samples are not actually the decoded frames, but rather just metadata, so it should be pretty efficient.

I thought about this some more and not sure what I was talking about before. We can step through the whole file without loading all samples into memory...

> Needs to be tested though, but I would like to support trimming empty frames from the end too as that's more common than empty frames at the start. And if we do trimming from the end, we could trim empty frames in the middle of the video too.

From what I've seen empty frames at the beginning of video files (i.e. simulator screen recordings) is caused by the start time metadata marker being negative / non zero (for example, it's `start_time=0.245000` for the example I posted (`ffprobe -i test.mp4 -show_format`)). Do you have an example of a file with blank frames at the end?

I apologize for the number of changes required; I know in the issue you requested that the PR be 99% done when opened.

**Comment by sindresorhus** [comment]
I think this one has one or more blank frames at the end: [Gifski.mp4.zip](https://github.com/sindresorhus/Gifski/files/6681948/Gifski.mp4.zip)

**Comment by codetheweb** [comment]
It looks fine to me in both QuickTime and the current version of Gifski:

Screen Shot 2021-06-20 at 1 30 58 PM

Screen Shot 2021-06-20 at 1 31 10 PM

**Comment by sindresorhus** [comment]
If you play it in Quick Look, it's empty at the end, but I guess, it's just that the video track ends earlier than the actual video, which Gifski already handles by rewrapping the video track into a fresh asset.

I'm sure some users are having problems with empty frames at the end as I have seen it in encode failure errors on Crashlytics, but I don't think I actually have a fixture for this.

Would you be able to make a fixture for this? Maybe take the video with the empty leading frames and move those frames to the end or something.

**Comment by codetheweb** [comment]
> Would you be able to make a fixture for this? Maybe take the video with the empty leading frames and move those frames to the end or something.

I would if I could, but as mentioned before as far as I can tell it's only possible to have blank / empty frames at the beginning of tracks, not the end (assuming the video track in a file gets moved into its own asset like Gifski does to prevent interference from audio tracks). It's easy to prepend / append a few frames of black to files with ffmpeg, but that's not the same thing as a blank frame.

**Comment by sindresorhus** [comment]
I think you are right. Let's go with only trimming at the start then.

File: Gifski/VideoValidator.swift [hunk 3]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
...
+					let currentTimestamp = CMSampleBufferGetOutputPresentationTimeStamp(sampleBuffer)
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 9455603.

File: Gifski/VideoValidator.swift [hunk 4]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
...
+
+					// Add video track to composition and remove time range (from the begining up to current frame)
+					let composition = AVMutableComposition()
+					let videoCompTrack = composition.addMutableTrack(withMediaType: AVMediaType.video, preferredTrackID: CMPersistentTrackID())
+
+					try videoCompTrack?.insertTimeRange(CMTimeRangeMake(start: .zero, duration: newAsset.duration), of: newAsset.firstVideoTrack!, at: .zero)
+					videoCompTrack?.removeTimeRange(CMTimeRange(start: .zero, end: currentTimestamp))
+
+					reader.cancelReading()
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 9455603.

File: Gifski/VideoValidator.swift [hunk 5]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
 			return .failure
 		}
 
-		return .success(newAsset, newVideoMetadata)
+		// Find first non-empty frame
+		do {
+			let reader = try AVAssetReader(asset: newAsset)
+			let readerOutput = AVAssetReaderTrackOutput(track: newAsset.firstVideoTrack!, outputSettings: nil)
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 9455603.

File: Gifski/VideoValidator.swift [hunk 6]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
...
+			while let sampleBuffer = readerOutput.copyNextSampleBuffer() {
+				// On first non-empty frame
+				if sampleBuffer.totalSampleSize != 0 {
+					let currentTimestamp = CMSampleBufferGetOutputPresentationTimeStamp(sampleBuffer)
+
+					// Add video track to composition and remove time range (from the begining up to current frame)
+					let composition = AVMutableComposition()
+					let videoCompTrack = composition.addMutableTrack(withMediaType: AVMediaType.video, preferredTrackID: CMPersistentTrackID())
+
+					try videoCompTrack?.insertTimeRange(CMTimeRangeMake(start: .zero, duration: newAsset.duration), of: newAsset.firstVideoTrack!, at: .zero)
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 9455603.

File: Gifski/VideoValidator.swift [hunk 7]
```diff
@@ -184,6 +184,47 @@ struct VideoValidator {
...
+			reader.startReading()
+
+			while let sampleBuffer = readerOutput.copyNextSampleBuffer() {
+				// On first non-empty frame
+				if sampleBuffer.totalSampleSize != 0 {
+					let currentTimestamp = CMSampleBufferGetOutputPresentationTimeStamp(sampleBuffer)
+
+					// Add video track to composition and remove time range (from the begining up to current frame)
+					let composition = AVMutableComposition()
+					let videoCompTrack = composition.addMutableTrack(withMediaType: AVMediaType.video, preferredTrackID: CMPersistentTrackID())
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 9455603.

File: Gifski/VideoValidator.swift (hunk 1 above)
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 9455603.

File: Gifski/Utilities.swift [hunk 8]
```diff
@@ -592,6 +592,54 @@ extension AVAsset {
 		imageGenerator.requestedTimeToleranceBefore = .zero
 		return imageGenerator.image(at: time)
 	}
+
+	func trimBlankFrames() -> AVAsset? {
+		guard let reader = try? AVAssetReader(asset: self) else {
+			return nil
```
Comment (codetheweb) [inline Gifski/Utilities.swift:598]: I followed the existing pattern of returning optionals; but would we want to throw in this extension instead?

Comment (sindresorhus) [reply Gifski/Utilities.swift:598]: Yes, the function should be throwing.

File: Gifski/Utilities.swift [hunk 9]
```diff
@@ -592,6 +592,54 @@ extension AVAsset {
...
+		reader.startReading()
+
+		while let sampleBuffer = readerOutput.copyNextSampleBuffer() {
+			if reader.status != .reading {
+				reader.cancelReading()
+				return nil
+			}
+
+			// On first non-empty frame
+			if sampleBuffer.totalSampleSize != 0 {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:616]: Use `guard`

File: Gifski/Utilities.swift [hunk 10]
```diff
@@ -592,6 +592,54 @@ extension AVAsset {
...
+		guard let videoTrack = self.firstVideoTrack else {
+			return nil
+		}
+
+		let readerOutput = AVAssetReaderTrackOutput(track: videoTrack, outputSettings: nil)
+		reader.add(readerOutput)
+		reader.startReading()
+
+		while let sampleBuffer = readerOutput.copyNextSampleBuffer() {
+			if reader.status != .reading {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:610]: Use `guard`

File: Gifski/Utilities.swift [hunk 11]
```diff
@@ -592,6 +592,54 @@ extension AVAsset {
...
+
+				// Add video track to composition and remove time range (from the begining up to current frame)
+				let composition = AVMutableComposition()
+				guard let videoCompTrack = composition.addMutableTrack(withMediaType: .video, preferredTrackID: CMPersistentTrackID()) else {
+					reader.cancelReading()
+					return nil
+				}
+
+				do {
+					try videoCompTrack.insertTimeRange(CMTimeRangeMake(start: .zero, duration: self.duration), of: videoTrack, at: .zero)
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:627]: Use the modern initializer, not `CMTimeRangeMake`.

File: Gifski/Utilities.swift [hunk 12]
```diff
@@ -592,6 +592,54 @@ extension AVAsset {
 		imageGenerator.requestedTimeToleranceBefore = .zero
 		return imageGenerator.image(at: time)
 	}
+
+	func trimBlankFrames() -> AVAsset? {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:596]: I think there should be a method on `AVAssetTrack` (with the same name) that contains the main logic and this method should just use it.

Comment (sindresorhus) [inline Gifski/Utilities.swift:596]: And this needs a doc comment.

File: Gifski/Utilities.swift (hunk 8 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in a3a64ed.

File: Gifski/Utilities.swift (hunk 9 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in a3a64ed.

File: Gifski/Utilities.swift (hunk 11 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in a3a64ed.

File: Gifski/Utilities.swift (hunk 12 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in a3a64ed.

Comment (codetheweb) [reply Gifski/Utilities.swift]: Makes sense, added in a3a64ed.

File: Gifski/Utilities.swift (hunk 10 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in a3a64ed.

File: Gifski/Utilities.swift [hunk 13]
```diff
@@ -584,6 +584,16 @@ extension AVAssetImageGenerator {
 }
 
 
+enum VideoTrimmingError: Error {
```
Comment (codetheweb) [inline Gifski/Utilities.swift:587]: Wasn't sure how specific / detailed you wanted the error types to be.

File: Gifski/Utilities.swift [hunk 14]
```diff
@@ -758,6 +787,63 @@ extension AVAssetTrack {
...
+				throw VideoTrimmingError.assetReaderFailure
+			}
+
+			// On first non-empty frame
+			guard sampleBuffer.totalSampleSize == 0 else {
+				let currentTimestamp = sampleBuffer.outputPresentationTimeStamp
+
+				// Add video track to composition and remove time range (from the begining up to current frame)
+				let composition = AVMutableComposition()
+				guard let videoCompTrack = composition.addMutableTrack(withMediaType: .video, preferredTrackID: CMPersistentTrackID()) else {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:819]: Not sure what the intention is with `CMPersistentTrackID()`? You could just write `0` if that's what you meant.

File: Gifski/Utilities.swift [hunk 15]
```diff
@@ -758,6 +787,63 @@ extension AVAssetTrack {
...
+					throw error
+				}
+
+				videoCompTrack.removeTimeRange(CMTimeRange(start: .zero, end: currentTimestamp))
+
+				reader.cancelReading()
+
+				guard let trimmedTrack = AVPlayerItem(asset: composition).asset.firstVideoTrack else {
+					throw VideoTrimmingError.compositionIsMissingTrack
+				}
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:837]: I'm not sure what's going on here. You could just return:

```swift
composition.asset.firstVideoTrack
```

File: Gifski/Utilities.swift [hunk 16]
```diff
@@ -758,6 +787,63 @@ extension AVAssetTrack {
 	}
 }
 
+extension AVAssetTrack {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:790]: I think both these extension should live together in the file.

File: Gifski/Utilities.swift [hunk 17]
```diff
@@ -758,6 +787,63 @@ extension AVAssetTrack {
...
+				do {
+					try videoCompTrack.insertTimeRange(.init(start: .zero, duration: asset.duration), of: self, at: .zero)
+				} catch {
+					reader.cancelReading()
+					throw error
+				}
+
+				videoCompTrack.removeTimeRange(CMTimeRange(start: .zero, end: currentTimestamp))
+
+				reader.cancelReading()
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:833]: I think you can just put this inside a `defer` at the top of the scope.

File: Gifski/Utilities.swift [hunk 18]
```diff
@@ -592,6 +602,25 @@ extension AVAsset {
...
 		return imageGenerator.image(at: time)
 	}
+
+	/**
+	Removes blank frames from the begining of the asset.
+
+	This can be useful to trim blank frames from files produced by tools like the iOS simulator screen recorder.
+	*/
+	func trimBlankFrames() throws -> AVAsset {
+		guard let videoTrack = self.firstVideoTrack else {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:612]:
```suggestion
		guard let videoTrack = firstVideoTrack else {
```

File: Gifski/Utilities.swift [hunk 19]
```diff
@@ -758,6 +787,63 @@ extension AVAssetTrack {
...
 }
 
+extension AVAssetTrack {
+	/**
+	Removes blank frames from the begining of the track.
+
+	This can be useful to trim blank frames from files produced by tools like the iOS simulator screen recorder.
+	*/
+	func trimBlankFrames() throws -> AVAssetTrack {
+		guard let asset = self.asset else {
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:797]:
```suggestion
		guard let asset = asset else {
```

File: Gifski/VideoValidator.swift [hunk 20]
```diff
@@ -184,6 +184,20 @@ struct VideoValidator {
...
-		return .success(newAsset, newVideoMetadata)
+		// Trim asset
+		do {
+			let trimmedAsset = try newAsset.trimBlankFrames()
+
+			return .success(trimmedAsset, newVideoMetadata)
+		} catch {
+			NSAlert.showModalAndReportToCrashlytics(
+				for: window,
+				title: "Could not trim empty frames from video.",
```
Comment (sindresorhus) [inline Gifski/VideoValidator.swift:195]:
```suggestion
				title: "Could not trim empty leading frames from video.",
```

File: Gifski/VideoValidator.swift [hunk 21]
```diff
@@ -184,6 +184,20 @@ struct VideoValidator {
 			return .failure
 		}
 
-		return .success(newAsset, newVideoMetadata)
+		// Trim asset
+		do {
+			let trimmedAsset = try newAsset.trimBlankFrames()
+
+			return .success(trimmedAsset, newVideoMetadata)
+		} catch {
```
Comment (sindresorhus) [inline Gifski/VideoValidator.swift:192]: You should not ignore the error. It should be included in the alert.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: Gifski/VideoValidator.swift (hunk 21 above)
Comment (codetheweb) [reply Gifski/VideoValidator.swift:192]: Fixed in 50532fc.

File: Gifski/VideoValidator.swift (hunk 20 above)
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Fixed in 50532fc.

File: Gifski/Utilities.swift (hunk 17 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 50532fc.

File: Gifski/Utilities.swift [hunk 22]
```diff
@@ -608,22 +598,87 @@ extension AVAsset {
...
-			throw VideoTrimmingError.assetIsMissingTrack
+			throw AVAssetTrack.VideoTrimmingError.assetIsMissingTrack
 		}
 
-		let trimmedTrack = try videoTrack.trimBlankFrames()
+		let trimmedTrack = try videoTrack.trimmingBlankFrames()
 
 		guard let trimmedAsset = trimmedTrack.asset else {
-			throw VideoTrimmingError.assetIsNil
+			throw AVAssetTrack.VideoTrimmingError.assetIsNil
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:609]: You can add a typealias for the error to `AVAsset`.

File: Gifski/Utilities.swift (hunk 19 above)
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/Utilities.swift]: Good suggestion, turned out much cleaner: 50532fc.

File: Gifski/Utilities.swift [hunk 23]
```diff
@@ -758,6 +787,63 @@ extension AVAssetTrack {
...
+		let reader = try AVAssetReader(asset: asset)
+
+		let readerOutput = AVAssetReaderTrackOutput(track: self, outputSettings: nil)
+		reader.add(readerOutput)
+		reader.startReading()
+
+		while let sampleBuffer = readerOutput.copyNextSampleBuffer() {
+			guard reader.status == .reading else {
+				reader.cancelReading()
+				throw VideoTrimmingError.assetReaderFailure
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 50532fc.

File: Gifski/Utilities.swift [hunk 24]
```diff
@@ -608,22 +598,87 @@ extension AVAsset {
...
+		throw VideoTrimmingError.assetIsEmpty
+	}
+}
+
+
+extension AVAssetTrack.VideoTrimmingError: LocalizedError {
+	public var errorDescription: String? {
+		self.rawValue
+	}
+}
```
Comment (sindresorhus) [inline Gifski/Utilities.swift]: This is an abuse of `LocalizedError`. You need to actually do a switch where you write human-readable messages for each error.

File: Gifski/Utilities.swift [hunk 25]
```diff
@@ -592,6 +592,90 @@ extension AVAsset {
...
+
+
+extension AVAssetTrack {
+	enum VideoTrimmingError: String, Error {
+		case assetIsNil
+		case assetReaderFailure
+		case assetIsEmpty
+		case assetIsMissingTrack
+		case compositionCouldNotBeCreated
+		case compositionIsMissingTrack
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:624]: This is unused.

File: Gifski/VideoValidator.swift [hunk 26]
```diff
@@ -184,6 +184,20 @@ struct VideoValidator {
...
+		// Trim asset
+		do {
+			let trimmedAsset = try newAsset.trimmingBlankFrames()
+
+			return .success(trimmedAsset, newVideoMetadata)
+		} catch {
+			NSAlert.showModalAndReportToCrashlytics(
+				for: window,
+				title: "Could not trim empty leading frames from video.",
+				message: "Error: \(error.localizedDescription). This should not happen. Email sindresorhus@gmail.com and include this info:",
```
Comment (sindresorhus) [inline Gifski/VideoValidator.swift:196]:
```suggestion
				message: "\(error.localizedDescription)\n\nThis should not happen. Email sindresorhus@gmail.com and include this info:",
```

File: Gifski/VideoValidator.swift [hunk 27]
```diff
@@ -184,6 +184,20 @@ struct VideoValidator {
 			return .failure
 		}
 
-		return .success(newAsset, newVideoMetadata)
+		// Trim asset
+		do {
+			let trimmedAsset = try newAsset.trimmingBlankFrames()
+
```
Comment (sindresorhus) [inline Gifski/VideoValidator.swift:190]:
```suggestion
```

File: Gifski/VideoValidator.swift (hunk 26 above)
Comment (codetheweb) [reply Gifski/VideoValidator.swift]: Example of how it looks now:
Screen Shot 2021-06-24 at 2 01 43 PM

File: Gifski/Utilities.swift (hunk 25 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 3a2c8a6.

File: Gifski/Utilities.swift (hunk 24 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 3a2c8a6.

File: Gifski/Utilities.swift (hunk 22 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Added in 3a2c8a6.

**Comment by codetheweb** [comment]
Thanks for all the detailed feedback; I really appreciate it.

File: Gifski/Utilities.swift [hunk 28]
```diff
@@ -595,6 +595,103 @@ extension AVAsset {
...
+	*/
+	func trimmingBlankFrames() throws -> AVAsset {
+		guard let videoTrack = self.firstVideoTrack else {
+			throw VideoTrimmingError.assetIsMissingTrack
+		}
+
+		let trimmedTrack = try videoTrack.trimmingBlankFrames()
+
+		guard let trimmedAsset = trimmedTrack.asset else {
+			throw VideoTrimmingError.assetIsNil
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:686]: Can this actually happen? I don't think so. I think it would be better to use an `assertionFailure` and just fall back to an empty `AVMutableComposition()`.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/Gifski/pull/245/files#r657856913 was marked as resolved, but was not resolved.

File: Gifski/Utilities.swift [hunk 29]
```diff
@@ -595,6 +595,103 @@ extension AVAsset {
...
+	Removes blank frames from the begining of the asset.
+
+	This can be useful to trim blank frames from files produced by tools like the iOS simulator screen recorder.
+	*/
+	func trimmingBlankFrames() throws -> AVAsset {
+		guard let videoTrack = self.firstVideoTrack else {
+			throw VideoTrimmingError.assetIsMissingTrack
+		}
+
+		let trimmedTrack = try videoTrack.trimmingBlankFrames()
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:683]: I just realized that this is kinda faulty. It works fine for Gifski, but it it drops any other video tracks and audio tracks. That is not expected from the method name and its docs.

File: Gifski/Utilities.swift [hunk 30]
```diff
@@ -595,6 +595,102 @@ extension AVAsset {
...
+
+			// On first non-empty frame.
+			guard sampleBuffer.totalSampleSize == 0 else {
+				let currentTimestamp = sampleBuffer.outputPresentationTimeStamp
+				wrappedTrack.removeTimeRange(.init(start: .zero, end: currentTimestamp))
+				return wrappedTrack
+			}
+		}
+
+		throw VideoTrimmingError.assetIsEmpty
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:647]: This error is misnamed. It should be `videoTrackIsEmpty`.

File: Gifski/Utilities.swift [hunk 31]
```diff
@@ -595,6 +595,102 @@ extension AVAsset {
 }
 
 
+extension AVAssetTrack {
+	enum VideoTrimmingError: Error {
+		case assetIsNil
+		case assetReaderFailure
+		case assetIsEmpty
+		case assetIsMissingTrack
```
Comment (sindresorhus) [inline Gifski/Utilities.swift:603]:
```suggestion
		case assetIsMissingVideoTrack
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: Gifski/Utilities.swift (hunk 29 above)
Comment (codetheweb) [reply Gifski/Utilities.swift:683]: It would be fairly straightforward to modify `AVAsset.trimmingBlankFrames()` so that it retained other tracks by creating a new composition; however video / audio might get out of sync then. Would you recommend just renaming it or modifying the logic?

Comment (sindresorhus) [reply Gifski/Utilities.swift:683]: Just renaming it.

File: Gifski/Utilities.swift [hunk 32]
```diff
@@ -595,6 +595,103 @@ extension AVAsset {
 }
 
 
+extension AVAssetTrack {
+	enum VideoTrimmingError: Error {
+		case assetIsNil
+		case assetReaderFailure
```
[the remark this answers is not in the document]
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 9a81320.

File: Gifski/Utilities.swift (hunk 28 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Probably right, fixed in 9a81320.

File: Gifski/Utilities.swift (hunk 29 above)
Comment (codetheweb) [reply Gifski/Utilities.swift:680]: Updated in 9a81320.

File: Gifski/Utilities.swift (hunk 30 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 9a81320.

File: Gifski/Utilities.swift (hunk 31 above)
Comment (codetheweb) [reply Gifski/Utilities.swift]: Fixed in 9a81320.

**Comment by sindresorhus** [comment]
Looks good now. Thanks :)
